### ISSUE TITLE
super-high cost when searching for documents

### ISSUE THREAD

**Issue by Roiocam** [apache/incubator-pekko#1097]
## Reproduce

Just search from https://pekko.apache.org/docs/pekko/current//index.html

<img width="1667" alt="截屏2024-02-04 00 43 51" src="https://github.com/apache/incubator-pekko/assets/26020358/301723b8-9a46-466c-af6d-65098ad4fd5c">

<img width="706" alt="截屏2024-02-04 00 49 34" src="https://github.com/apache/incubator-pekko/assets/26020358/cfea3c1d-6460-49ce-bd48-30dd7804f72e">

<img width="611" alt="截屏2024-02-04 00 50 09" src="https://github.com/apache/incubator-pekko/assets/26020358/20f7c77b-da19-4908-bc26-8803b7cab4df">

## Performance Profile

[Trace-20240204T004355.json](https://github.com/apache/incubator-pekko/files/14152202/Trace-20240204T004355.json)

**Comment by Roiocam** [comment]
After some digging, I just found the issue caused by a very large search_index.json, it will load when the user clicks the search input box.


<img width="930" alt="截屏2024-02-04 02 34 49" src="https://github.com/apache/incubator-pekko/assets/26020358/1c33d9e5-315c-419c-b14f-d408dddf2400">

This javascript file comes from the upstream repo: https://github.com/squidfunk/mkdocs-material, in the latest documentation website of them, it will pre-loaded search index, and the index file size is only 200kb.

Then I checked the content of pekko search index file, which is very large json array.

<img width="815" alt="截屏2024-02-04 02 48 11" src="https://github.com/apache/incubator-pekko/assets/26020358/a2d1451a-0765-4505-a5e3-46ac3a7655a6">

Finally, I decided to look up the answer in the upstream repo, and I found this: https://github.com/squidfunk/mkdocs-material/issues/904

**Comment by pjfanning** [comment]
It is Paradox that builds this json file. I'm not sure what we can do in the short term. Maybe, we should look at trying to offload the search to Google instead of using our own search with its own JSON file.

**Comment by pjfanning** [comment]
An example Google search

`actor site:pekko.apache.org`

https://www.google.com/search?q=actor+site%3Apekko.apache.org

**Comment by He-Pin** [comment]
IIRC, it was using the algolia for indexing and searching

**Comment by Roiocam** [comment]
> IIRC, it was using the algolia for indexing and searching

I don't think so, only akka uses algolia, we use paradox-material-theme which depends on  mkdocs-material

>  Maybe, we should look at trying to offload the search to Google instead of using our own search with its own JSON file.

I am using Docusaurus at work, they use the same way to implement offline search but won't stuck UI thread.

**Comment by Roiocam** [comment]
I will try to handle this issue via an update paradox-material-theme.

**Comment by mdedetrich** [comment]
> I will try to handle this issue via an update paradox-material-theme.

Note that [sbt-paradox-material-theme](https://github.com/sbt/sbt-paradox-material-theme) was just transferred to the sbt org/community and we are currently in the process of making the necessary changes so it may take a bit of time before we can get to deploying the change

**Comment by Roiocam** [comment]
> I will try to handle this issue via an update paradox-material-theme.

I tried to upgrade to the latest mkdocs-material and found that the way it was built changed after the 5.x version, and only up-to 5.x version can avoid search blocking the main thread, and it seems that the implementation is through the need to precompile a search index file. 

I think this upgrade is no less difficult than rewriting upstream https://github.com/sbt/sbt-paradox-material-theme. We should consider replacing the search implementation to solve this ISSUE.

**Comment by squidfunk** [comment]
Author of Material for MkDocs here. v5.x is from 2020, so pretty old. We made [significant improvements on search](https://squidfunk.github.io/mkdocs-material/blog/2021/09/13/search-better-faster-smaller/) in [9.x](https://squidfunk.github.io/mkdocs-material/changelog/#9.0.0), which should be twice as fast and significantly cut down on index size, and will be replacing it with an entirely new implementation that will be much faster and more powerful in the near future. Related:

- https://github.com/squidfunk/mkdocs-material/issues/6307

**Comment by mdedetrich** [comment]
@Roiocam sbt-paradox-theme has been ported over to the sbt org/package and 0.7.0 has just been [published](https://github.com/sbt/sbt-paradox-material-theme/actions/runs/7901103220/job/21564226923) so you are now free to make changes against [sbt-paradox-theme](https://github.com/sbt/sbt-paradox-material-theme).
